kernel/src/path.rs

```diff
@@ -1,34 +1,108 @@
 //! Utilities to make working with directory and file paths easier
 
-use lazy_static::lazy_static;
-use regex::Regex;
 use url::Url;
 
 use crate::{DeltaResult, Version};
 
-/// The delimiter to separate object namespaces, creating a directory structure.
+/// The delimiter to separate object namespaces, creating a directory structure. Note this is in url
+/// terms, so we use `/`
 const DELIMITER: &str = "/";
 
-lazy_static! {
-    static ref CHECKPOINT_FILE_PATTERN: Regex =
-        Regex::new(r"\d+\.checkpoint(\.\d+\.\d+)?\.parquet").unwrap();
-    static ref DELTA_FILE_PATTERN: Regex = Regex::new(r"\d+\.json").unwrap();
+#[derive(Debug)]
+pub(crate) struct LogPath<'a> {
+    url: &'a Url,
+    path: &'a str,
+    version: Option<Version>,
+    // if is compacted, this path spans version [`version`, `compacted_to_version`]
+    _compacted_to_version: Option<Version>,
+    is_commit: bool,
+    is_checkpoint: bool,
 }
 
-#[derive(Debug)]
-pub(crate) struct LogPath<'a>(pub(crate) &'a Url);
+fn get_filename(path: &str) -> Option<&str> {
+    if path.is_empty() || path.ends_with('/') {
+        None
+    } else {
+        path.rsplit(DELIMITER).next()
+    }
+}
+
+fn get_version_opt(version_str_opt: Option<&str>, expected_digits: usize) -> Option<Version> {
+    version_str_opt.and_then(|version_str| {
+        if version_str.len() == expected_digits {
+            version_str.parse().ok()
+        } else {
+            None
+        }
+    })
+}
+
+pub(crate) fn version_from_location(location: &Url) -> Option<Version> {
+    let path = location.path();
+    get_filename(path)
+        .and_then(|f| f.split_once('.'))
+        .and_then(|(name, _)| get_version_opt(Some(name), 20))
+}
 
 impl<'a> LogPath<'a> {
+    pub(crate) fn new(url: &'a Url) -> Self {
+        let path = url.path();
+        let version_str = get_filename(path).and_then(|f| f.split_once('.'));
+        let version = version_str.and_then(|(name, _)| get_version_opt(Some(name), 20));
+
+        let mut is_commit = false;
+        let mut is_checkpoint = false;
+        let mut compacted_to_version = None;
+        if version.is_some() {
+            // could be a checkpoint or commit file, let's check
+            let (_, suffix) = version_str.unwrap(); // safe, version.is_some()
+            is_commit = suffix == "json"; // if we were just [version].json, we're a commit file
+
+            if !is_commit && suffix.starts_with("checkpoint.") {
+                let rest = &suffix[11..];
```

Comment:
Same as above, perhaps a constant?

Reply:
This is just the length of the string we checked for right above. I've just added a comment to explain taht.